File: Assets/AboutXLua/Scripts/Framework/Dialogue/CsharpOnly/DialogueFuncRegistry.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;
using XLua;

// C#版本对话系统函数注册与调用器
public static class DialogueFuncRegistry
{
    private static Dictionary<string, MethodInfo> _funcMap = new();

    /// <summary>
    /// 扫描实现了IDialogueFuncProvider接口的类
    /// </summary>
    public static void ScanAndRegister()
    {
        _funcMap.Clear();
        var assembly = Assembly.GetExecutingAssembly();
        
        // 只获取实现了IDialogueFuncProvider接口的类型
        var targetTypes = assembly.GetTypes()
            .Where(t => typeof(IDialogueFuncProvider).IsAssignableFrom(t) && !t.IsInterface);

        foreach (var type in targetTypes)
        {
            // 只处理静态方法
            var methods = type.GetMethods(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
            foreach (var method in methods)
            {
                var attr = method.GetCustomAttribute<DialogueFuncAttribute>();
                if (attr != null)
                {
                    string funcName = !string.IsNullOrEmpty(attr.DisplayName) ? 
                        attr.DisplayName : method.Name;
                    if (string.IsNullOrEmpty(funcName))
                    {
                        Debug.LogWarning($"[DialogueFuncRegistry] 对话函数名不能为空：{type.Name}.{method.Name}");
                        continue;
                    }

                    if (_funcMap.ContainsKey(funcName))
                    {
                        Debug.LogWarning($"[DialogueFuncRegistry] 对话函数名重复：{funcName}（{type.Name}.{method.Name}）");
                        continue;
                    }

                    _funcMap.Add(funcName, method);
                    Debug.Log($"[DialogueFuncRegistry] 注册对话函数：{funcName} -> {type.Name}.{method.Name}");
                }
            }
        }
    }

    /// <summary>
    /// 调用对话函数
    /// </summary>
    public static object InvokeFunction(string funcName, params object[] parameters)
    {
        _funcMap.TryGetValue(funcName,out var method);
        if (method == null)
        {
            Debug.LogError($"[DialogueFuncRegistry] 未找到对话函数：{funcName}");
            return null;
        }

        try
        {
            // 参数适配：根据目标方法的参数类型修正推导的参数类型
            var adaptedParams = AdaptParameters(method, parameters);
            return method.Invoke(null, adaptedParams);
        }
        catch (Exception e)
        {
            Debug.LogError($"[DialogueFuncRegistry] 执行对话函数出错 {funcName}：{e.Message}");
            return null;
        }
    }

    /// <summary>
    /// 参数适配：根据目标方法的参数类型签名修正推导的参数类型
    /// 解决自动类型推导导致的类型不匹配问题（如期望string的"123"被推导成double）
    /// </summary>
    private static object[] AdaptParameters(MethodInfo method, object[] parameters)
    {
        var paramInfos = method.GetParameters();
        if (parameters == null || parameters.Length == 0)
            return parameters;

        var adapted = new object[parameters.Length];
        for (int i = 0; i < parameters.Length; i++)
        {
            var value = parameters[i];
            // 如果有对应的参数信息，则尝试适配类型
            if (i < paramInfos.Length)
            {
                var expectedType = paramInfos[i].ParameterType;
                adapted[i] = AdaptValue(value, expectedType);
            }
            else
            {
                // 超出定义的参数，保持原样（可能是params参数）
                adapted[i] = value;
            }
        }
        return adapted;
    }

    /// <summary>
    /// 将值适配为期望类型
    /// </summary>
    private static object AdaptValue(object value, Type expectedType)
    {
        if (value == null)
            return null;

        var valueType = value.GetType();
        
        // 如果类型已经匹配，直接返回
        if (expectedType.IsAssignableFrom(valueType))
            return value;

        try
        {
            // 期望string类型，将任意类型转换为字符串
            if (expectedType == typeof(string))
            {
                return value.ToString();
            }
            
            // 期望数值类型，尝试转换
            if (expectedType == typeof(int) || expectedType == typeof(Int32))
            {
                return Convert.ToInt32(value);
            }
            if (expectedType == typeof(long) || expectedType == typeof(Int64))
            {
                return Convert.ToInt64(value);
            }
            if (expectedType == typeof(float) || expectedType == typeof(Single))
            {
                return Convert.ToSingle(value);
            }
            if (expectedType == typeof(double) || expectedType == typeof(Double))
            {
                return Convert.ToDouble(value);
            }
            if (expectedType == typeof(bool) || expectedType == typeof(Boolean))
            {
                // 字符串转bool
                if (value is string strVal)
                {
                    var lower = strVal.ToLower();
                    if (lower == "true" || lower == "1") return true;
                    if (lower == "false" || lower == "0") return false;
                }
                return Convert.ToBoolean(value);
            }

            // 处理List<object>类型
            if (expectedType == typeof(List<object>) && value is Dictionary<object, object> dict)
            {
                // 如果是连续整数键的字典，转为列表
                return DictToList(dict);
            }
            
            // 特殊处理：List<object> 转 List<string> (解决泛型转换失败 InvalidCastException)
            if (expectedType == typeof(List<string>) && value is List<object> objList)
            {
                return objList.Select(o => o?.ToString()).ToList();
            }

            // 特殊处理：String 转 List (防止解析失败导致单个字符串传给List参数)
            if (typeof(System.Collections.IList).IsAssignableFrom(expectedType) && value is string sVal)
            {
                if (expectedType == typeof(List<string>))
                    return new List<string> { sVal };
                if (expectedType == typeof(List<object>))
                    return new List<object> { sVal };
            }

            // 使用ChangeType作为兜底
            return Convert.ChangeType(value, expectedType);
        }
        catch (Exception e)
        {
            Debug.LogWarning($"[DialogueFuncRegistry] 参数类型适配失败：{value}({valueType.Name}) -> {expectedType.Name}，使用原值。错误：{e.Message}");
            return value;
        }
    }

    /// <summary>
    /// 将连续整数键的字典转换为列表
    /// </summary>
    private static List<object> DictToList(Dictionary<object, object> dict)
    {
        var list = new List<object>();
        int index = 1;
        while (dict.TryGetValue((double)index, out var val) || dict.TryGetValue(index, out val))
        {
            list.Add(val);
            index++;
        }
        return list.Count > 0 ? list : dict.Values.ToList();
    }
}

File: Assets/Dialogue/LuaScripts/DialogueSystem/DialogueFuncRegistry.lua
---@class DialogueFuncRegistry 对话函数注册
local DialogueFuncRegistry = {}

--- 模拟接口标记模块，用于扫描约束
DialogueFuncRegistry.IDialogueFuncProvider = {}

--- 函数注册表：函数名 -> 函数
--- funcName = {func, module, paramTypes}
--- paramTypes 为可选的参数类型数组，用于参数适配：{"string", "number", "boolean", "table"}
local _funcMap = {}

---@function 将值适配为期望类型
---@param value any 待适配的值
---@param expectedType string 期望类型："string", "number", "boolean", "table", nil表示不转换
local function AdaptValue(value, expectedType)
    if value == nil or expectedType == nil then
        return value
    end

    local currentType = type(value)
    
    -- 如果类型已匹配，直接返回
    if currentType == expectedType then
        return value
    end

    -- 期望string类型，将任意类型转换为字符串
    if expectedType == "string" then
        return tostring(value)
    end
    
    -- 期望number类型，尝试转换
    if expectedType == "number" then
        if currentType == "string" then
            return tonumber(value) or value
        end
        return tonumber(value) or value
    end
    
    -- 期望boolean类型
    if expectedType == "boolean" then
        if currentType == "string" then
            local lower = string.lower(value)
            if lower == "true" or lower == "1" then return true end
            if lower == "false" or lower == "0" then return false end
        end
        if currentType == "number" then
            return value ~= 0
        end
        return value and true or false
    end
    
    -- 期望table类型，暂不转换
    if expectedType == "table" then
        return value
    end
    
    return value
end

---@function 参数适配：根据期望的参数类型列表修正推导的参数类型
---@param params table 参数列表
---@param paramTypes table|nil 期望的参数类型列表
local function AdaptParameters(params, paramTypes)
    if not params or not paramTypes then
        return params
    end
    
    local adapted = {}
    for i, value in ipairs(params) do
        local expectedType = paramTypes[i]
        adapted[i] = AdaptValue(value, expectedType)
    end
    return adapted
end

---@function 扫描实现 IDialogueFuncProvider 的模块
---@param module table
---@param moduleName string
function DialogueFuncRegistry.ScanModule(module, moduleName)
    if not IsImplementProvider(module) then
        CS.UnityEngine.Debug.LogWarning(("[Lua] 模块 %s 未实现 IDialogueFuncProvider 接口，跳过扫描"):format(moduleName))
        return
    end

    for key, item in pairs(module) do
        -- 跳过元字段（__开头）和非表类型
        if type(key) == "string" and not key:find("^__") and type(item) == "table" then
            -- 严格校验：必须含 name(string) + func(function)
            if type(item.name) == "string" and type(item.func) == "function" then
                local funcName = item.name
                if _funcMap[funcName] then
                    CS.UnityEngine.Debug.LogWarning(("[Lua] 函数名冲突: %s (原模块:%s, 新模块:%s)"):format(
                            funcName, _funcMap[funcName].module, moduleName))
                else
                    -- paramTypes 为可选的参数类型数组
                    _funcMap[funcName] = { 
                        func = item.func, 
                        module = moduleName,
                        paramTypes = item.paramTypes  -- 可选：期望的参数类型列表
                    }
                    CS.UnityEngine.Debug.Log(("[Lua] 注册函数: %s (模块:%s)"):format(funcName, moduleName))
                end
            end
        end
    end
end

---@function 调用Lua对话函数
---@param funcName string 函数名
---@param params table 入参列表
function DialogueFuncRegistry.InvokeFunction(funcName, params)
    local funcInfo = _funcMap[funcName]
    if not funcInfo then
        CS.UnityEngine.Debug.LogError("[Lua-DialogueFuncRegistry] 未找到Lua对话函数："..funcName)
        return nil
    end
    
    params = params or {}
    -- 参数适配：根据期望的参数类型修正推导的参数类型
    local adaptedParams = AdaptParameters(params, funcInfo.paramTypes)
    
    local status, result = pcall(funcInfo.func, table.unpack(adaptedParams))
    if not status then
        CS.UnityEngine.Debug.LogError("[Lua-DialogueFuncRegistry] 执行Lua函数"..funcName.."出错:"..result)
        return nil
    end
    return result
end

---@function 取消注册对话函数
---@param moduleName string 要取消注册的模块名
function DialogueFuncRegistry.UnregisterFunction(moduleName)
    for funcName, funcInfo in pairs(_funcMap) do
        if funcInfo.module == moduleName then
            _funcMap[funcName] = nil
        end
    end
end

---@function 检查模块是否实现IDialogueFuncProvider接口（元表判断）
---@param module table 待检查的Lua模块 
function IsImplementProvider(module)
    local mt = getmetatable(module)
    return mt and mt.__index == DialogueFuncRegistry.IDialogueFuncProvider
end

return DialogueFuncRegistry